staging/src/k8s.io/client-go/features/testing/features.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package testing
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"testing"
+
+	clientfeatures "k8s.io/client-go/features"
+)
+
+var (
+	overrideLock                   sync.Mutex
+	currentlyOverridingTestName    string
+	originalGatesForOverridingTest clientfeatures.Gates
+)
+
+// SetFeatureGatesDuringTest sets the specified gate for the given feature to the specified value for the duration of the test.
+//
+// Example use:
+//
+//	clientfeaturestesting.SetFeatureGatesDuringTest(t, clientfeatures.FeatureGates(), clientfeatures.WatchListClient, false)
+func SetFeatureGatesDuringTest(tb testing.TB, gate clientfeatures.Gates, f clientfeatures.Feature, newValue bool) {
```

Comment:
This is not how I understood @liggitt's comment. @liggitt , I would appreciate it if you could clarify whether you were against replacing the implementation or if you wanted to ensure that the method could be used by a test that might read multiple features.

I think that avoiding replacing the implementation means changing the production code to facilitate testing. My intention here was not to change the production code or even test the default implementation. I want to have a method that allows me to test the code behind a feature flag.

Reply:
The ReplaceFeatureGates() functionality looks problematic.

As a small example, anything that called `features.FeatureGates()` previously would now be holding the wrong instance. That seems very likely to cause init / ordering problems.

I don't think the godoc / warning will be sufficient to keep people from getting themselves in trouble with ReplaceFeatureGates() or this implementation.

> I think that avoiding replacing the implementation means changing the production code to facilitate testing.

Having the production code facilitate testing doesn't seem bad to me...

> I want to have a method that allows me to test the code behind a feature flag.

Agreed that is necessary, does the pattern in the existing SetFeatureGateDuringTest not work?

Reply:
> As a small example, anything that called features.FeatureGates() previously would now be holding the wrong instance. That seems very likely to cause init / ordering problems.

This scenario came up in https://github.com/kubernetes/kubernetes/pull/121012#discussion_r1348735489 as well, does returning a proxy object from `FeatureGates()` (that always proxies to the current effective implementation) address this?

Separately, do we need to allow feature gate overrides from multiple places to be in effect at the same time? I am thinking of two cases:

1. A test overrides one or more feature gates for the duration of the test.
2. An application process (e.g. a kube component) sets overrides early on in its life, in `main()` or at least before it starts to run gated Kubernetes code, and those overrides remain in place until the process dies.

Is it programmer error (panic-worthy) if a caller tries to set a replacement with a replacement already in effect?

Reply:
Personally, I prefer the solution with proxy because we will be able to leave the default implementation unchanged. Currently, the default implementation only reads environment variables once and sets the state. It would be nice if it stayed that way.

As for the second question, I think that tests should signal (panic) a change in feature from various tests. This would allow detecting tests that would be otherwise unstable. I don't see the need to add a similar mechanism to the production code.

Reply:
@liggitt would you be okay with the proxy idea? I would like to get your confirmation before implementing it. Thanks.

Reply:
> Does returning a proxy object from `FeatureGates()` (that always proxies to the current effective implementation) address this?

I think so, yes

> do we need to allow feature gate overrides from multiple places to be in effect at the same time?

I'm not sure about multiple overrides of distinct gates from different places for tests.

We definitely *don't* want multiple concurrent overrides of the same gate from different places for tests.

We definitely *do* want a test to only have to override the gates it cares about and have the non overridden gates fall through to their non-overridden state.

Reply:
OK, pushed, I still need to add a few things but it shows the direction.

Reply:
Alright, I've added the code I wanted, PTAL

Reply:
@liggitt friendly ping to review the PR. Thanks!

Reply:
the signature here looks much better, thanks

will defer to @deads2k's review of the rest